addons/website_sale_loyalty/models/loyalty.py

```diff
@@ -0,0 +1,75 @@
+    _name = 'loyalty.program'
+    _inherit = ['loyalty.program']
+
+    def get_available_rewards(self, order):
+        """ Returns the list of rewards that the cart can get """
+        client = order.partner_id
+        rewards = []
+        if client:
+            spendable_points = tools.float_round(client.loyalty_points - order.spent_loyalty_points, 0, rounding_method='HALF-UP')
+            for reward in self.reward_ids:
+                if reward.minimum_points > spendable_points:
+                    continue
+                elif reward.reward_type == 'discount' and reward.point_cost > spendable_points:
+                    continue
+                elif reward.reward_type == 'gift' and reward.point_cost > spendable_points:
+                    continue
+                elif reward.reward_type == 'discount' and reward.discount_apply_on == 'specific_products':
+                    found = False
+                    for line in order.order_line:
+                        found |= line.product_id in reward.discount_specific_product_ids
+                    if not found:
+                        continue
+                elif reward.reward_type == 'discount' and reward.discount_type == 'fixed_amount' and order.amount_total < reward.minimum_amount:
+                    continue
+                already_in_cart = False
+                for line in order.order_line:
+                    if line.loyalty_reward_id and line.loyalty_reward_id.id == reward.id:
+                        already_in_cart = True
+                if already_in_cart:
+                    continue
+                rewards.append(reward)
+        return rewards
+
+    def contains_discount_reward(self, product_id):
+        """Returns true if this reward contains the specified product_id as a discount reward"""
+        for reward in self.reward_ids:
+            if reward.reward_type == 'discount' and reward.discount_product_id.sudo().product_tmpl_id.id == product_id.id:
+                return True
+        return False
+
+
+class LoyaltyRule(models.Model):
+    _name = 'loyalty.rule'
+    _inherit = ['loyalty.rule']
+
+    valid_product_ids = fields.One2many('product.product', compute='_compute_valid_product_ids')
+
+    @api.depends('rule_domain')
+    def _compute_valid_product_ids(self):
+        for rule in self:
+            if rule.rule_domain:
+                domain = safe_eval(rule.rule_domain)
+                rule.valid_product_ids = self.env['product.product'].search(domain)
+            else:
+                rule.valid_product_ids = self.env['product.product'].search([])
+
+    def is_product_valid(self, product_id):
+        """Avoid fetching the full product list if no domain is defined"""
+        if self.rule_domain:
+            return product_id in self.valid_product_ids
+        else:
+            return True
+
+
+class LoyaltyReward(models.Model):
+    _name = 'loyalty.reward'
+    _inherit = ['loyalty.reward', 'image.mixin']
```

Comment:
Why not in `loyalty` module?

Reply:
It will be strange if images can be defined for rewards in POS but are not used when displaying those rewards

Reply:
Yes I do agree and I thought about it.
My idea was that if another module is inheriting `loyalty` module, there is lot of chances they will also need the image (unlike the POS which is special case where no client is directly involved).
So it would be weird to have another module having to also define the `image.mixin`.
On the other hand, it would mean image on loyalty reward not used in the module the mixin is inherited, a bit strange too.
But it is probably premature optimization, so let's forget about it.